### PR TITLE
Preprocess api file in devops pipeline for Python

### PR DESCRIPTION
**Pull request by praveenkuttappan**
Added new APIs to accept requests to generate review using published preprocessed API file. Existing APIs are not removed for backward compatibility for other languages.

### CODE REVIEW

File: src/dotnet/APIView/APIViewWeb/Controllers/AutoReviewController.cs [hunk 1]
```diff
@@ -67,5 +67,28 @@ public async Task<ActionResult> GetReviewStatus(string language, string packageN
 
             throw new Exception("Automatic review is not found for package " + packageName);
         }
+
+        [HttpGet]
+        public async Task<ActionResult> createApiReview(
```
Comment (weshaggard) [inline src/dotnet/APIView/APIViewWeb/Controllers/AutoReviewController.cs:72]:
```suggestion
        public async Task<ActionResult> CreateApiReview(
```
nit: .NET usually capitalizes method names.

File: src/dotnet/APIView/APIViewWeb/Controllers/PullRequestController.cs [hunk 2]
```diff
@@ -38,7 +38,27 @@ public PullRequestController(PullRequestManager pullRequestManager, ILogger<Auto
                 return StatusCode(StatusCodes.Status400BadRequest);
             }
 
-            await _pullRequestManager.DetectApiChanges(buildId, artifactName, filePath, pullRequestNumber, commitSha, repoName, packageName);
+            await _pullRequestManager.DetectApiChanges(buildId, artifactName, filePath, commitSha, repoName, packageName, pullRequestNumber);
+            return Ok();
+        }
+
+        [HttpGet]
+        public async Task<ActionResult> createApiReview(
```
Comment (weshaggard) [inline src/dotnet/APIView/APIViewWeb/Controllers/PullRequestController.cs:46]:
```suggestion
        public async Task<ActionResult> CreateApiReview(
```

File: src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs [hunk 3]
```diff
@@ -510,6 +520,54 @@ public async void UpdateReviewBackground()
                     _telemetryClient.StopOperation(operation);
                 }
             }
+        }
+
+        public async Task<CodeFile> GetCodeFile(string repoName, string buildId, string artifactName, string packageName, string originalFileName, string codeFileName, MemoryStream memoryStream)
+        {
+            Stream stream = null;
+            CodeFile codeFile = null;
+            if (string.IsNullOrEmpty(codeFileName))
```
Comment (weshaggard) [inline src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs:529]: What cases do we want to only download and extract a single file?

Comment (weshaggard) [reply src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs:529]: If there is a scenario then I believe the download api might take a file pattern you can pass, at least the download task does in DevOps.

Comment (praveenkuttappan) [reply src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs:529]: This is for backward compatibility to support API reviews for PR until we switch over all languages to new sand boxing model. Conditionally allowing either "current" file or new zip allow us to migrate language by language to new model.

Comment (weshaggard) [reply src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs:529]: Oh I missed the fact this isn't extracting a file from a zip it is just downloading a single file.

File: src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs [hunk 4]
```diff
@@ -510,6 +520,54 @@ public async void UpdateReviewBackground()
...
+            {
+                // backward compatibility until all languages moved to sandboxing of codefile to pipeline
+                stream = await _devopsArtifactRepository.DownloadPackageArtifact(repoName, buildId, artifactName, originalFileName, "file");
+                codeFile = await CreateCodeFile(Path.GetFileName(originalFileName), stream, false, memoryStream);
+            }
+            else
+            {
+                stream = await _devopsArtifactRepository.DownloadPackageArtifact(repoName, buildId, artifactName, packageName, "zip");
```
Comment (weshaggard) [inline src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs:537]: Will we require the file already be a zip or will devops already do that if we point it to an artifact?

**Review by weshaggard — APPROVED** [review]
Looks reasonable.

File: src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs (hunk 4 above)
Comment (praveenkuttappan) [reply src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs:537]: devops does it for us when we request download. We don't need to zip them in advance

**Pull request merged**